File: JavaScript/159-Longest Substring with At Most Two Distinct Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstringTwoDistinct = function(s) {
    let hmap = new Map();
    let left = 0;
    let res = 0;
    
    for (let i=0; i<s.length; i++){
        hmap.set(s[i], (hmap.get(s[i])||0)+1 );
        while (hmap.size > 2){
            hmap.set(s[left], hmap.get(s[left])-1);
            if (hmap.get(s[left]) === 0) hmap.delete(s[left]);
            left++;
        }
        res = Math.max(res, i-left+1);
    }
    return res;
};

File: C++/159-Longest Substring with At Most Two Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        unordered_map<char, int> count;
        int res = 0;
        int left = 0;
        for (int i=0; i<s.size(); i++){
            count[s[i]]++;
            while (count.size() > 2){
                count[s[left]]--;
                if (count[s[left]] == 0) count.erase(s[left]);
                left++;
            }
            res = max(res, i-left+1);
        }
        return res;
    }
};